index.bs

```diff
@@ -677,14 +692,515 @@ id=] must be the same as the [=window handle=].
+protocol will result in an error.
+
+A [=/session=] has an <dfn>object id map</dfn>. This is a weak map from objects to
+their corresponding id.
+
+Issue: Should this be explicitly per realm?
+
+<div algorithm>
+To get the <dfn export>object id for an object</dfn> given an |object|:
+
+  1. If [=object id map=] for the [=current session=] does not contain |object|
+     run the following steps:
+
+     1. Let |object id| be a new, unique, string identifier for |object|. If
+        |object| is an [=/element=] this must be the [=web element reference=]
+        for |object|; if it's [=WindowProxy=] object, this must be the
+        [=window handle=] for |object|.
+
+    1. Set the value of |object| in the [=object id map=] to |object id|.
+
+  1. Return the result of getting the value for |object| in [=object id map=].
+</div>
+
+<div algorithm>
+To get the <dfn export>object for an object id</dfn> given an |object id|:
+
+  1. For each |object| → |id value| of [=object id map=]:
+
+    1. If |id value| is equal to |object id|, return [=success=] with data
+       |object|
+
+  1. Return [=error=] with [=error code=] [=no such element=]
+
+  Issue: This error code isn't right.
+</div>
+
+[=remote end definition=] and [=local end definition=]
+```
+RemoteObject = {
+  UndefinedObject //
+  NullObject //
+  StringObject //
+  NumberObject //
+  BooleanObject //
+  BigIntObject //
+  SymbolObject //
+  ArrayObject //
+  FunctionObject //
+  ObjectObject //
+  FunctionObject //
+  RegexObject //
+  DateObject //
+  MapObject //
+  SetObject //
+  WeakMapObject //
+  WeakSetObject //
+  IteratorObject //
+  GeneratorObject //
+  ErrorObject //
+  ProxyObject //
+  PromiseObject //
+  TypedArrayObject //
+  ArrayBufferObject //
+  NodeObject //
+  WindowProxyObject //
+}
+
+ObjectId = text;
+
+UndefinedObject = {
+  type: "undefined",
+  value: null,
+}
+
+NullObject = {
+  type: "null",
+  value: null,
+}
+
+StringObject = {
+  type: "string",
+  value: text,
+}
+
+SpecialNumber = "NaN" / "-0" / "+Infinity" / "-Infinity";
+
+NumberObject = {
+  type: "number",
+  value: number / SpecialNumber,
+}
+
+BooleanObject = {
+  type: "boolean",
+  value: bool,
+}
+
+BigIntObject = {
+  type: "bigint",
+  value: text,
+}
+
+SymbolObject = {
+  type: "symbol",
+  objectId: ObjectId,
+}
+
+ArrayObject = {
+  type: "array",
+  objectId: ObjectId,
+  value?: [RemoteObject]
+}
+
+ObjectObject = {
+  type: "object",
+  objectId: ObjectId,
+  value?: {*text: RemoteObject}
+}
+
+FunctionObject = {
+  type: "function",
+  objectId: ObjectId,
+}
+
+RegexObject = {
+  type: "regex",
+  objectId: ObjectId,
+  value: text
+}
+
+DateObject = {
+  type: "date",
+  objectId: ObjectId,
+  value: text
+}
+
+MapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+SetObject = {
+  type: "set",
+  objectId: ObjectId
+}
+
+WeakMapObject = {
+  type: "map",
+  objectId: ObjectId,
+}
+
+WeakSetObject = {
+  type: "set",
+  objectId: ObjectId,
+}
+
+IteratorObject = {
```

Comment:
What JS value would end up being serialized as `IteratorObject`? `Symbol.iterator` it a Symbol and the actual iterators functions, so I'm not sure.

If there is something, how about async iterators?

Reply:
I assumed this was for things like `someIterable[Symbol.iterator]()`, i.e. objects that implement [the `Iterator` interface](https://tc39.es/ecma262/#sec-iterator-interface). Is that what you had in mind, @jgraham?

In that case, @foolip’s suggestion to include something like [`AsyncIteratorObject`](https://tc39.es/ecma262/#sec-asynciterator-interface) makes sense.

Reply:
If that is the intention, then something could be both an iterable and something else at the same time. It's more like a behavior of an object than a type of object.

Reply:
I think I"m going to pull out these interface types for now. I assume the intent here is that you can write a client which grabs an iterable remote object and provides an API to actually iterate over it, going back to the remote end to actually perform that operation. But I don't think we need to get this right on the first pass.